x/pylons/keeper/msg_get_tx_history.go

```diff
@@ -0,0 +1,306 @@
+	// adding records to collection
+	userHistory = append(userHistory, QueryEventNFTSell(history.TxResponses)...)
+
+	sort.Slice(userHistory, func(i, j int) bool {
+		return userHistory[i].CreatedAt >= userHistory[j].CreatedAt
+	})
+
+	// if incoming request is for a specific denomination
+	if len(denom) > types.MinVal {
+		his := []*types.History{}
+		// filter out the specific denomination from the tx history
+		for _, h := range userHistory {
+			if strings.Contains(h.Amount, denom) {
+				his = append(his, h)
+			}
+		}
+		userHistory = his
+	}
+
+	// calculating limit and offset to retutrn paginated slice of complete tx history
+	offset = limit * (offset - 1) // subtracted 1 as indices start at 0, case offset is 1 i.e. first page, where first index is 0
+	limit += offset
+
+	// case: if offset if outof bound the return empty response
+	if offset > int64(len(userHistory)) {
+		return []*types.History{}, nil
+	} else if limit > int64(len(userHistory)) { // case if limit is outof bounds the return history from offset to end
+		return userHistory[offset:], nil
+	}
+	// normal case pagination lies between the slices
+	return userHistory[offset:limit], nil
+}
+
+/*
+* Event information is contained with in the log entry of the TxResponse
+* In log entry events can be further found at log[n].Events
+* Hence, we first loop over all the txResponses in the history from cosmos
+* Then find our desired event and store it
+ */
+
+func QueryEventSender(block []*sdkTypes.TxResponse) (userHistory []*types.History) {
+	for _, txRes := range block {
+		// getting block time to note the time event occcured
+		// parsing date in string to date object
+		date, _ := time.Parse(time.RFC3339, txRes.Timestamp)
+		// using date to get unix timestamp of the block
+		// as we have 2 types from this block SEND and NFTBUY
+		entry := &types.History{
+			CreatedAt: date.Unix(),
+			Type:      types.TxTypeSend,
+		}
+		nft := &types.History{
+			CreatedAt: date.Unix(),
+			Type:      types.TxTypeNFTBuy,
+		}
+		for _, log := range txRes.Logs {
+		eventFound:
+			for _, e := range log.Events {
+				switch e.Type {
+				// case transfer event found, i.e. user have sent amount user bank send
+				case types.TransferEventKey:
+					for _, attr := range e.GetAttributes() {
+						// extract amount of transfer and the recipient address
+						switch attr.Key {
+						case types.KeyAmount:
+							if len(attr.Value) == types.MinVal {
+								entry.Amount = types.ValZero
+							} else {
+								entry.Amount = attr.Value
+							}
+						case types.KeyRecipient:
+							entry.Address = attr.Value
+						}
+					}
+					// case create_item event found extracyt amount at which nft is bought
+					// its identification id's and owner address
+				case types.CreateItemKey:
+					entry.Address = ""
+					for _, attr := range e.GetAttributes() {
+						switch attr.Key {
+						case types.KeyAmount:
+							if len(attr.Value) == types.MinVal {
```

Comment:
this if/else is multiple places within different functions, code reusability would be better.